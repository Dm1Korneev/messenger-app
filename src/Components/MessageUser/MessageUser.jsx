import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import ListItem from '@material-ui/core/ListItem';
import Typography from '@material-ui/core/Typography';

import UsersAvatar from 'Components/UsersAvatar';

const useStyles = makeStyles((theme) => ({
  ListItemText: {
    backgroundColor: theme.palette.background.default,
    zIndex: 999,
    padding: theme.spacing(0, 2),
  },
}));

const MessageUser = (props) => {
  const classes = useStyles();

  const {
    isCurrentUserMessage, author, avatar, children,
  } = props;

  const flexDirection = isCurrentUserMessage ? 'row-reverse' : 'row';

  return (
    <ListItem>
      <Box
        display="flex"
        flex="1"
        alignItems="start"
        flexDirection={flexDirection}
      >
        <Box position="sticky" top={0}>
          <UsersAvatar author={author} avatar={avatar} />
        </Box>
        <Box display="flex" flex="1" flexDirection="column">
          <Box
            position="sticky"
            top={0}
            display="flex"
            flexDirection={flexDirection}
            className={classes.ListItemText}
          >
            <Typography
              component="span"
              variant="subtitle2"
            >
              {author}
            </Typography>
          </Box>
          {children}
        </Box>
      </Box>
    </ListItem>
  );
};

MessageUser.propTypes = {
  isCurrentUserMessage: PropTypes.bool.isRequired,
  author: PropTypes.string.isRequired,
  avatar: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
};

export default MessageUser;

