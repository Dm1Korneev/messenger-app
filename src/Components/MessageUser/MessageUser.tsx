import Box from '@material-ui/core/Box';
import ListItem from '@material-ui/core/ListItem';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import { FC } from 'react';

import { UsersAvatar } from 'Components/UsersAvatar';

const useStyles = makeStyles((theme) => ({
  ListItemText: {
    backgroundColor: theme.palette.background.default,
    zIndex: 999,
  },
}));

type Props = {
  isCurrentUserMessage: boolean;
  author: string;
  avatar: string;
}

export const MessageUser: FC<Props> = ({
  isCurrentUserMessage, author, avatar, children,
}) => {
  const classes = useStyles();

  const flexDirection = isCurrentUserMessage ? 'row-reverse' : 'row';

  return (
    <ListItem>
      <Box
        display="flex"
        flex="1"
        alignItems="start"
        flexDirection={flexDirection}
      >
        <Box position="sticky" top={0}>
          <UsersAvatar author={author} avatar={avatar} />
        </Box>
        <Box display="flex" flex="1" flexDirection="column" px={2}>
          <Box
            position="sticky"
            top={0}
            display="flex"
            flexDirection={flexDirection}
            className={classes.ListItemText}
          >
            <Typography
              component="span"
              variant="subtitle2"
            >
              {author}
            </Typography>
          </Box>
          {children}
        </Box>
      </Box>
    </ListItem>
  );
};
