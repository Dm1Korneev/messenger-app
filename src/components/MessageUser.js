import React from "react";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import withStyles from "@material-ui/core/styles/withStyles";
import Typography from "@material-ui/core/Typography";
import classNames from "classnames";
import UsersAvatar from "./UsersAvatar";

function MessageUser(props) {
  const { isCurrentUserMessage, author, avatar, children, classes } = props;

  const ListItemTextClass = classNames(
    classes.ListItemText,
    isCurrentUserMessage && classes.reverse
  );

  const ListItemClass = classNames(
    classes.ListItem,
    isCurrentUserMessage && classes.reverse
  );

  return (
    <ListItem className={ListItemClass}>
      <div className={classes.sticky}>
        <UsersAvatar author={author} avatar={avatar} />
      </div>
      <div className={classes.textContent}>
        <ListItemText
          className={classNames(ListItemTextClass, classes.sticky)}
          primary={
            <Typography
              component="span"
              className={classes.ListItemText__author}
            >
              {author}
            </Typography>
          }
        />
        {children}
      </div>
    </ListItem>
  );
}

const styles = theme => ({
  ListItemText: {
    backgroundColor: theme.palette.background.default,
    zIndex: 999,
    display: "flex",
    justifyContent: "space-between",
    "&:first-child": {
      paddingLeft: theme.spacing.unit * 2
    }
  },
  ListItemText_reverse: {
    display: "flex",
    flexDirection: "row-reverse",
    justifyContent: "space-between"
  },
  reverse: {
    flexDirection: "row-reverse"
  },
  ListItem: {
    alignItems: "start",
    paddingTop: theme.spacing.unit,
    paddingBottom: 0,
    paddingLeft: theme.spacing.unit * 2
  },
  textContent: {
    flexGrow: 1
  },
  ListItemText__author: {
    fontSize: theme.typography.pxToRem(12),
    fontWeight: 700
  },
  sticky: {
    position: "sticky",
    top: 0
  }
});

export default withStyles(styles)(MessageUser);
