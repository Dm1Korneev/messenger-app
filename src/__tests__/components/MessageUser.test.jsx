import React from 'react';

import MessageUser from 'Components/MessageUser';

jest.mock('Components/UsersAvatar', () => global.mockComponent('UsersAvatar'));

jest.mock('@material-ui/core/ListItem', () => global.mockComponent('ListItem'));
jest.mock('@material-ui/core/ListItemText', () => global.mockComponent('ListItemText'));
jest.mock('@material-ui/core/Typography', () => global.mockComponent('Typography'));

const scrollIntoViewMock = jest.fn();
window.HTMLElement.prototype.scrollIntoView = scrollIntoViewMock;

const props = {
  isCurrentUserMessage: false,
  author: '',
  avatar: '',
  children: <></>,
};

describe('render MessageUser component', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = global.mount(
      <MessageUser {...props} />,
    );
  });

  test('component is render', () => {
    expect(wrapper.find('MessageUser').length).toBe(1);
  });

  test('ListItem subcomponent is render', () => {
    expect(wrapper.find({ originalcomponent: 'ListItem' }).length).toBe(1);
  });
});

describe('snapshot-test MessageUser component', () => {
  test('Renders correct properties', () => {
    global.mountExpect(
      <MessageUser {...props} />,
    ).toMatchSnapshot();
  });
});
