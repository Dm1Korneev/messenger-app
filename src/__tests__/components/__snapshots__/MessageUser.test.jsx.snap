// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot-test MessageUser component Renders correct properties 1`] = `
<WithStyles(MessageUser)
  author=""
  avatar=""
  isCurrentUserMessage={false}
>
  <MessageUser
    author=""
    avatar=""
    classes={
      Object {
        "ListItem": "MessageUser-ListItem-4",
        "ListItemText": "MessageUser-ListItemText-1",
        "ListItemText__author": "MessageUser-ListItemText__author-6",
        "ListItemText_reverse": "MessageUser-ListItemText_reverse-2",
        "reverse": "MessageUser-reverse-3",
        "sticky": "MessageUser-sticky-7",
        "textContent": "MessageUser-textContent-5",
      }
    }
    isCurrentUserMessage={false}
  >
    <Component
      className="MessageUser-ListItem-4"
    >
      <div
        originalcomponent="ListItem"
        testprops={
          Object {
            "children": Array [
              <div
                className="MessageUser-sticky-7"
              >
                <Unknown
                  author=""
                  avatar=""
                />
              </div>,
              <div
                className="MessageUser-textContent-5"
              >
                <Unknown
                  className="MessageUser-ListItemText-1 MessageUser-sticky-7"
                  primary={
                    <Unknown
                      className="MessageUser-ListItemText__author-6"
                      component="span"
                    >
                      
                    </Unknown>
                  }
                />
                <React.Fragment />
              </div>,
            ],
            "className": "MessageUser-ListItem-4",
          }
        }
      >
        <div
          className="MessageUser-sticky-7"
        >
          <Component
            author=""
            avatar=""
          >
            <div
              originalcomponent="UsersAvatar"
              testprops={
                Object {
                  "author": "",
                  "avatar": "",
                }
              }
            />
          </Component>
        </div>
        <div
          className="MessageUser-textContent-5"
        >
          <Component
            className="MessageUser-ListItemText-1 MessageUser-sticky-7"
            primary={
              <Unknown
                className="MessageUser-ListItemText__author-6"
                component="span"
              >
                
              </Unknown>
            }
          >
            <div
              originalcomponent="ListItemText"
              testprops={
                Object {
                  "className": "MessageUser-ListItemText-1 MessageUser-sticky-7",
                  "primary": <Unknown
                    className="MessageUser-ListItemText__author-6"
                    component="span"
                  >
                    
                  </Unknown>,
                }
              }
            />
          </Component>
        </div>
      </div>
    </Component>
  </MessageUser>
</WithStyles(MessageUser)>
`;
